### ISSUE TITLE
Inappropriate exception from BlockBlobClient.Upload()

### ISSUE THREAD

**Issue by peterchase** [Azure/azure-sdk-for-net#15787]
**Occurs When**

Using Azure.Storage.Blobs 12.6.0

Calling BlockBlobClient.Upload() with a 250MB MemoryStream with a slow-ish network that cannot upload that much data before timeout.

**Behaviour**

Upload() threw an AggregateException containing 6 TaskCancelledException instances (presumably because of retry). Caller had not cancelled any CancellationToken. None of the exceptions had a message giving any insight into the root cause.

**Expected**

Public API should not throw an OperationCancelledException (of which TaskCancelledException is a subclass) unless it is because a CancellationToken passed in by the caller has been cancelled - not the case here.

Public API should avoid throwing AggregateException unless genuinely unavoidable, as it is hard for the caller to handle.

The only exception documented for BlockBlobClient.Upload() is RequestFailedException. Ideally, such an exception should be thrown. If not, something like a TimeoutException would at least make the cause clear and be easy to handle separately from other exceptions that might need different treatment.

Exception message should clearly indicate what happened.

**Note**

No StorageTransferOptions were passed in when this happened, so the BlockBlobClient attempted to send the 250MB as a single PUT.

We are now passing appropriate StorageTransferOptions to avoid this particular instance of the inappropriate exception, but the poor exception could still be encountered, e.g. even slower network.

**Comment by jsquire** [comment]
//cc: @tg-msft

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team best able to assist.

**Comment by amnguye** [comment]
I would have expected the AggregateException to contain 6 OperationCancelledException instead of 6 TaskCancelledException, because TaskCancelledException would indicate that a CancellationToken was passed (which is not the case).

If you are experiencing network issues and the requests that are being sent are reaching the timeout, then OperationCancelledException is the correct exception to see. Seeing that it's 6 exceptions within AggregateException, it's safe to assume it is due to the timeout.

A `RequestFailedException` should indicate a failed HTTP error code (400+), so it might be best not to throw this type of exception. I like the idea of throwing a `TimeOutException` instead though, it's more fitting since the request failed due to network issues, and not storage service issues.

I did some digging and looks like this is a common `HttpClient` issue https://github.com/dotnet/runtime/issues/21965 with dotnet  2.1 and looks like they were looking to change it to `TimeOutException` in dotnet 5.0. We actually recently upgraded to dotnet 5.0 so I will to look to writing a specific test for dotnet 5.0 and see if we're throwing the correct `TimeOutException`, instead of the `TaskCancelledException`.

Thanks for submitting this issue.
